update to zw100, and update z-wave driver

caseda opened the pull request:
Driver.js:
- update report parser binary sensor to crashless version
- Add tamper alarm (see note below for extra info)
- Add GETs to the multilevel sensors
- Simplify parameter 41
- Add parameter 81 (device firmware 1.8 and up only)
- Simplify parameter 201 and remove parser, doing manual signed, in a
signed parameter
- remove parser parameter 202, doing manual signed, in a signed
parameter
- remove parser parameter 203, doing manual signed, in a signed
parameter
- remove parser parameter 204, doing manual signed, in a signed
parameter
- added a custom tamper cancellation timer (see note below for more
info)

config/zw100.json:
- Add tamper capability
- cleaned up the labels of settings, moving most information to the hint
- Add tamper cancellation setting (see note below for more info)
- Add parameter 81 (device firmware 1.8 and up only)
- Add value 0 (Nothing) to parameters 102 and 103, = default value, had
no dropdown value.

Extra information Tamper Capability:
The device only sends out an "Tamper Alarm!" notification when there is
being tampered with the device.
There is no "idle" value given when Getting the value, it only sends the
"Tamper Alarm!" notification back turning the tamper alarm on, the
reason for manually inserting a "false" value.
not the most reliable way, but better then null or always active until
the user tampers with the device him/herself.
The device also never sends a "cancel tamper" notification, so added a
manual cancellation in the settings to get it working consistenly in
homey for all devices.

Z-wave driver update was needed for the tamper capability to work
properly, so is mandatory!
PS: that async module was enormous though!

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -52,42 +53,42 @@
 			"hint": {
-				"en": "Time before motion alarm is turned off since last motion detected event, time unit is second if the value range is 10 - 255, if the range is 256 - 3600 the time unit will be minute and its value should be dividable by 60 without remainder",
-				"nl": "Tijd voordat bewegingsdetectie alarm wordt uitgeschakeld na laatste beweging, tijdseenheid is seconden als de waarde tussen 10 en 255 is, als de waarde tussen 256 en 3600 is is de tijdseenheid minuten en dan moet de waarde deelbaar zijn door 60 zonder restgetal"
+				"en": "Time before motion alarm is turned off since last motion detected event.\nRange: 10 - 255 seconds\n256 (/ 60) - 3600 (/ 60) minutes",
```
(/ 60) does not really make sense, (divided by 60) is more clear in my opinion.

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -52,42 +53,42 @@
-				"en": "Time before motion alarm is turned off since last motion detected event, time unit is second if the value range is 10 - 255, if the range is 256 - 3600 the time unit will be minute and its value should be dividable by 60 without remainder",
-				"nl": "Tijd voordat bewegingsdetectie alarm wordt uitgeschakeld na laatste beweging, tijdseenheid is seconden als de waarde tussen 10 en 255 is, als de waarde tussen 256 en 3600 is is de tijdseenheid minuten en dan moet de waarde deelbaar zijn door 60 zonder restgetal"
+				"en": "Time before motion alarm is turned off since last motion detected event.\nRange: 10 - 255 seconds\n256 (/ 60) - 3600 (/ 60) minutes",
+				"nl": "Tijd voordat bewegingsdetectie alarm wordt uitgeschakeld na laatste beweging.\nBereik: 10 - 255 seconden\n256 (/ 60) - 3600 (/ 60) minuten"
```
same ^

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -152,17 +174,25 @@
-				"en": "Automatic reporting when value threshold reached",
-				"nl": "Automatisch rapportern wanneer waarde grens heeft overschreden"
+				"en": "Threshold Dependend Reports",
+				"nl": "Drempel Afhankelijke Verzendingen"
```
'Verzendingen'?:p I'd argue for 'Updates' or something similar.

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -152,17 +174,25 @@
+			},
+			"hint": {
+				"en": "If temperature threshold (Celsius) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 10 degree",
+				"nl": "Temperatuur verandering (Celsius) nodig voordat het apparaat een nieuwe waarde verstuurd op het interval\nBereik: 0 (uit), 1 - 10 graden"
```
verstuurt*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -174,8 +204,12 @@
-				"en": "Humidity threshold value, if exceeded device will send an automatic report, range: 0 (off), 1 - 100 %",
-				"nl": "Vochtigheidsgraad verandering nodig voordat het apparaat een automatische update verstuurt, bereik: 0 (uit), 1 - 100 %"
+				"en": "Humidity Threshold",
+				"nl": "Lucht Vochtigheids Drempel"
```
Luchtvochtigheid*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -174,8 +204,12 @@
+				"nl": "Lucht Vochtigheids Drempel"
+			},
+			"hint": {
+				"en": "If humidity threshold (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
```
percentage should not be capitalized

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -174,8 +204,12 @@
+			},
+			"hint": {
+				"en": "If humidity threshold (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
+				"nl": "Lucht vochtigheids verandering (Percentage) nodig voordat het apparaat een nieuwe waarde verstuurd op het interval\nBereik: 0 (uit), 1 - 100%"
```
Luchtvochtigheid verandering*
percentage should not be capitalized

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -187,8 +221,12 @@
-				"en": "Luminance threshold value, if exceeded device will send an automatic report, range: 0 (off), 1 - 1000 lux",
-				"nl": "Licht verandering nodig voordat het apparaat een automatische update verstuurt, bereik: 0 (uit), 1 - 1000 lux"
+				"en": "Luminance Threshold",
+				"nl": "Omgevings Licht Drempel"
```
Omgevingslicht*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -187,8 +221,12 @@
+			},
+			"hint": {
+				"en": "If luminance threshold (Lux) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 1000 lux",
+				"nl": "Omgevings licht verandering (Lux) nodig voordat het apparaat een nieuwe waarde verstuurd op het interval\nBereik: 0 (uit), 1 - 1000 lux"
```
Omgevingslicht*
verstuurt*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -200,8 +238,12 @@
-				"en": "Battery threshold value, if exceeded device will send an automatic report, range: 0 (off), 1 - 100 %",
-				"nl": "Batterij lading verandering nodig voordat het apparaat een automatische update verstuurt, bereik: 0 (uit), 1 - 100 %"
+				"en": "Battery Level Threshold",
+				"nl": "Batterij Niveau Drempel"
```
Batterijniveau*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -200,8 +238,12 @@
+				"nl": "Batterij Niveau Drempel"
+			},
+			"hint": {
+				"en": "If battery level (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
```
Percentage should not be capitalized

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -200,8 +238,12 @@
+			},
+			"hint": {
+				"en": "If battery level (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
+				"nl": "Batterij niveau verandering (Percentage) nodig voordat het apparaat een nieuwe waarde verstuurd op het interval\nBereik: 0 (uit), 1 - 100%"
```
Batterijniveau*
Percentage should not be capitalized
verstuurt*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -213,8 +255,12 @@
+				"nl": "Ultraviolet Drempel"
+			},
+			"hint": {
+				"en": "If ultraviolet (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
```
Percentage should not be capitalized

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -213,8 +255,12 @@
+			},
+			"hint": {
+				"en": "If ultraviolet (Percentage) has been exceeded, the device will send the new value on the interval\nRange: 0 (off), 1 - 100%",
+				"nl": "Ultraviolet verandering (Percentage) nodig voordat het apparaat een nieuwe waarde verstuurd op het interval\nBereik: 0 (uit), 1 - 100%"
```
Percentage should not be capitalized
verstuurt*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -223,19 +269,39 @@
+			"type": "checkbox",
+			"label": {
+				"en": "LED Behaviour On Alarms",
+				"nl": "LED Gedrag Met Alarmen"
```
LED Gedrag Bij Alarm, sounds better to me

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -223,19 +269,39 @@
+				"nl": "LED Gedrag Met Alarmen"
+			},
+			"hint": {
+				"en": "Enable the LED on motion and tamper alarms.\n\nFirmware v1.8 and up",
```
Maybe make it more clear it is about the device's firmware, not Homey's.

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -223,19 +269,39 @@
+			},
+			"hint": {
+				"en": "Enable the LED on motion and tamper alarms.\n\nFirmware v1.8 and up",
+				"nl": "Activeer de LED bij beweging en sabotage alarmen.\n\nFirmware v1.8 en hoger"
```
See comment above

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -340,12 +413,12 @@
-				"en": "Reporting interval association group 1",
-				"nl": "Informatie update interval associatie groep 1"
+				"en": "Report Interval Association Group 1",
+				"nl": "Verzend Interval Associatie Groep 1"
```
Again, maybe 'Update' is more descriptive?

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -340,12 +413,12 @@
 			"hint": {
-				"en": "This parameter indicates how often the Multi Sensor 6 should send an update with new information to devices in association group 1",
-				"nl": "Deze parameter geeft aan hoe vaak de Multi Sensor 6 een update met nieuwe informatie naar apparaten in de eerste associatie group zou moeten sturen"
+				"en": "This parameter indicates how often the Multi Sensor 6 should send an update with new information to devices in association group 1 (Homey).",
```
You could remove the last part of the sentence: '.. should send an update with new information to Homey'

caseda replied:
There can be multiple devices (max 5) in association group 1.
So i don't think that is the right sentence then.

RobinBol replied:
Ah ok, I guessed it was controller only!

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -340,12 +413,12 @@
-				"en": "This parameter indicates how often the Multi Sensor 6 should send an update with new information to devices in association group 1",
-				"nl": "Deze parameter geeft aan hoe vaak de Multi Sensor 6 een update met nieuwe informatie naar apparaten in de eerste associatie group zou moeten sturen"
+				"en": "This parameter indicates how often the Multi Sensor 6 should send an update with new information to devices in association group 1 (Homey).",
+				"nl": "Deze parameter geeft aan hoe vaak de Multi Sensor 6 een update met nieuwe informatie naar apparaten in de eerste associatie group zou moeten sturen (Homey)"
```
See comment above

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -357,8 +430,8 @@
-				"en": "Reporting interval association group 2",
-				"nl": "Informatie update interval associatie groep 2"
+				"en": "Report Interval Association Group 2",
+				"nl": "Verzend Interval Associatie Groep 2"
```
Maybe 'Update'?

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -374,8 +447,8 @@
-				"en": "Reporting interval association group 3",
-				"nl": "Informatie update interval associatie groep 3"
+				"en": "Report Interval Association Group 3",
+				"nl": "Verzend Interval Associatie Groep 3"
```
Maybe 'Update'?

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -391,12 +464,12 @@
 			"hint": {
-				"en": "Determine how much degree celsius needs to be offset to get the right temperature, range: -12.8 - +12.7 (Sensor Firmware v1.7 and up)",
-				"nl": "Bepaal hoeveel graden celsius de temperatuur gecompenseerd moet worden, bereik: -12,8 - +12,7 (Sensor Firmware v1.7 en hoger)"
+				"en": "Determine how much degree celsius needs to be offset to get the right temperature.\nRange: -12.8 - +12.7\n\nFirmware v1.7 and up",
```
Maybe make it more clear it is about the device's firmware

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -409,12 +482,12 @@
-				"en": "Humidity calibration value",
-				"nl": "Vochtigheidsgraad kalibratie waarde"
+				"en": "Humidity Calibration",
+				"nl": "Lucht Vochtigheid Kalibratie"
```
Luchtvochtigheid*

RobinBol commented on config/drivers/ZW100.json:
```diff
@@ -426,12 +499,12 @@
-				"en": "Luminance calibration value",
-				"nl": "Licht kalibratie waarde"
+				"en": "Luminance Calibration",
+				"nl": "Omgevings Licht Kalibratie"
```
Omgevingslicht*

RobinBol commented on drivers/ZW100/driver.js:
```diff
@@ -159,50 +199,30 @@ module.exports = new ZwaveDriver(path.basename(__dirname), {
-				value = Math.round(value * 10);
-
-				// If value is negative, subtract value from 256
-				if (value < 0) value = 256 + value;
```
Why did you remove this step? Seems crucial to me?

caseda replied:
I've explained this in my PR description:
The value is already being parsed as signed in the zwave driver, and this method does exactly the same as the signed buffer translation, but then more savely checked by nodejs.
Doing it "twice" makes no sense and is just more prone to errors/mistakes in the driver.

It was a change of the periode I didn't know about the signed values.

RobinBol replied:
Ok, makes sense, that was not clear to me from the PR description.

RobinBol commented on drivers/ZW100/driver.js:
```diff
@@ -159,50 +199,30 @@ module.exports = new ZwaveDriver(path.basename(__dirname), {
 			size: 1,
-			parser: value => {
-				// If value is negative, subtract value from 256
-				if (value < 0) value = 256 + value;
```
Again, why removed?

RobinBol commented on drivers/ZW100/driver.js:
```diff
@@ -159,50 +199,30 @@ module.exports = new ZwaveDriver(path.basename(__dirname), {
 			size: 2,
-			parser: value => {
-				// If value is negative, subtract value from 65536
-				if (value < 0) value = 65536 + value;
```
Again, why removed?

RobinBol commented on drivers/ZW100/driver.js:
```diff
@@ -159,50 +199,30 @@ module.exports = new ZwaveDriver(path.basename(__dirname), {
-			},
+		},
+		tamper_cancellation: (newValue, oldValue, deviceData) => {
+			if (tamperCancellation.hasOwnProperty(deviceData.token) && tamperCancellation[deviceData.token]) {
```
Why do you need to reset the alarm_tamper when a user saves a new tamper_cancellation period? Is the value automatically saved in node.settings.tamper_cancellation (I did not use these kind of settings before)?

caseda replied:
It only resets the timeout function, so if the user puts it lower, it will go off sooner like the user expects to happen. If higher it will take longer.

caseda replied:
no, had that wrong idea in my mind 😛 (has been a while when i wrote this parser)
i did it so it starts with a clean slate, have though about this one some time,

1: if it would only clear the timeout the alarm will stay on until the user triggers the tamper manually,
2: if it did nothing and the user had a big value before, then go to a low value it will stay on that entire long periode.
or the other way around, when the user has a low value, then go to the biggest value, it will go off early before the first manual trigger, maybe leaving the user confused

caseda commented:
Seems like I need a better spell checker :joy:

caseda commented:
ok... and then i pressed enter too soon :joy:
i've fixed all typo's (hopefully) and removed all parameters that were meant for association groups 2 and 3, since they don't exist.
also bumped default value for report interval to 600 seconds according to possible data flooding <10 minute interval in device firmware 1.08
